PR to review #4173

levsnv opened the pull request:
https://github.com/rapidsai/cuml/pull/4173

levsnv committed:
a sketch

levsnv committed:
Merge branch 'branch-21.08' of github.com:rapidsai/cuml into HEAD

levsnv committed:
added inference code

levsnv committed:
Merge branch 'branch-21.08' of github.com:rapidsai/cuml into HEAD

levsnv committed:
missed some errors

levsnv committed:
now compiles

levsnv committed:
stash

levsnv committed:
implemented omp reduction for memory pool size estimation

levsnv committed:
Revert "stash"

This reverts commit 472575da199ae30457f231ab97b00a0ac7a72185.

levsnv committed:
enums are more printable now

levsnv committed:
fixed some bugs; added debug prints

levsnv committed:
Merge branch 'branch-21.08' of github.com:rapidsai/cuml into categorical

levsnv committed:
stack trace

levsnv committed:
added allocation/copy code

levsnv committed:
implemented filling up the bit pool with category masks

levsnv committed:
fixed a type mismatch, added assert

levsnv committed:
drafting categorical node and set generation in test code

levsnv committed:
fixed compiler but not runtime issues

levsnv committed:
many things

cpu inference supports categorical nodes
data generation supports categorical features
moved is_categoricals_h generation from GPU to CPU (value flow more obvious)
fitted parameters to conventions
neater memory management

levsnv committed:
debugging

levsnv committed:
fixed some bugs

levsnv committed:
...

levsnv committed:
fixed some bugs

levsnv committed:
resolved merge conflicts, to restore lost code chunks

levsnv committed:
fixed all but a missing void base_node::print()

levsnv committed:
added void base_node::print()

levsnv committed:
Merge branch 'categorical-no-import' into categorical-treelite

levsnv committed:
fixed some compiler messsages

levsnv committed:
...

levsnv committed:
Merge branch 'categorical-no-import' into categorical-treelite

levsnv committed:
fixed a couple of bugs - operator precedence and swap def_left on export and import

levsnv committed:
fixed bugs related to treelite tests, commented extra prints

levsnv committed:
Treelite GTIL has different logic for margin vs two-sided probabilities vs class label prediction, disable when not needed

levsnv committed:
some review comments

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

trxcllnt committed:
Remove -g from cython compile commands (#4172)

Removes `-g` from the compile commands generated by distutils to compile Cython files.

This will make our container images, conda packages, and python wheels smaller.

levsnv committed:
Merge remote-tracking branch '10.2/categorical-no-import' into categorical-treelite

levsnv committed:
Merge remote-tracking branch '10.2/categorical-no-import' into categorical-treelite

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-no-import

levsnv committed:
Merge

levsnv committed:
Merge remote-tracking branch '10.2/categorical-no-import' into categorical-treelite

levsnv committed:
style

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of /home/ldolgovs/others-libs/rapids/host-envs/cuml-10.2 into categorical-treelite

levsnv committed:
re-fixed a bug?

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

levsnv committed:
add lightgbm tests

levsnv committed:
tried optimizing by working on full matrices at once, ended up pessimizing and un-categorizing the features

levsnv committed:
now supporting intermixed numerical and categorical features

levsnv committed:
merged into existing test

levsnv committed:
cleaned up diff

levsnv committed:
python style

levsnv committed:
added cat_sets to model shape string

levsnv committed:
removed prints, style

levsnv committed:
stray change

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

levsnv committed:
removed *_repr, gtil checks

levsnv committed:
removed unused variables and a debug getenv()

levsnv committed:
fix printing of max_matching -1

levsnv committed:
Merge branch 'categorical-no-import' into categorical-treelite

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

levsnv committed:
Merge branch 'categorical-no-import' of github.com:levsnv/cuml into categorical-treelite

levsnv committed:
BITS_PER_BYTE; comment

levsnv committed:
style, odd code

dantegd committed:
Pytest updates for Scikit-learn 0.24 (#4205)



Authors:
  - Dante Gama Dessavre (https://github.com/dantegd)

Approvers:
  - Victor Lafargue (https://github.com/viclafargue)
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4205

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -52,7 +52,7 @@ std::ostream& operator<<(std::ostream& os, const cat_sets_owner& cso)
   os << "\nbits { ";
   for (uint8_t b : cso.bits) {
-    os << std::bitset<8>(b) << " ";
+    os << std::bitset<BITS_PER_BYTE>(b) << " ";
```
This looks like it belongs to a different pull request.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
 
-inline void adjust_threshold(
-  float* pthreshold, int* tl_left, int* tl_right, bool* default_left, tl::Operator comparison_op)
+cat_feature_counters reduce(cat_feature_counters a, cat_feature_counters b)
```
`reduce` is a very generic name. For `cat_feature_counters`, a more specific and descriptive name would be better.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+          .n_nodes      = a.n_nodes + b.n_nodes};
+}
+
+std::vector<cat_feature_counters> reduce(std::vector<cat_feature_counters> a,
```
Again, a more descriptive name would be useful.

canonizer replied:
What is the intent of this function?

- If it performs a pairwise reduction of vector elements and returns the result (like `+`), the signature should be `std::vector<T> reduce(const std::vector<T>& a, const std::vector<T>& b)`.

- If it pairwise reduces the elements of the 2nd vector into the 1st vector (like `+=`), it should be `std::vector<T>& reduce_into(std::vector<T>& a, const std::vector<T>& b)`.

In either case, the function signature should be changed.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+}
+
+std::vector<cat_feature_counters> reduce(std::vector<cat_feature_counters> a,
+                                         const std::vector<cat_feature_counters> b)
```
```suggestion
                                         const std::vector<cat_feature_counters>& b)
```

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+}
+
+template <typename T, typename L>
+inline std::vector<cat_feature_counters> cat_features_counters(const tl::Tree<T, L>& tree,
```
Does this function extract a list of category feature counters from a tree? In any case, a documentation comment would be useful.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical &&
+          tree.HasMatchingCategories(node_id)) {
+        std::vector<uint32_t> matching_cats = tree.MatchingCategories(node_id);
+        uint32_t max_matching_cat           = *(matching_cats.end() - 1);
```
```suggestion
        uint32_t max_matching_cat           = matching_cats.back();
```

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+               "more than %d matching categories",
+               MAX_PRECISE_INT_FLOAT);
+        cat_feature_counters& counters = res[tree.SplitIndex(node_id)];
+        counters                       = reduce(counters, {(int)max_matching_cat, 1});
```
Why is a cast needed here?

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+
+// modifies cat_sets
+template <typename fil_node_t, typename T, typename L>
+conversion_state<fil_node_t> tl2fil_branch_node(int fil_left_child,
```
`tl2fil_inner_node`

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -625,23 +626,92 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+  return cat_features;
+}
+
+void adjust_threshold(float* pthreshold,
```
Please only use this function in treelite-to-FIL import.

If you need to perform similar transformations to generate treelite models in the tests, please use a separate function for that.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+    if (tree.CategoriesListRightChild(tl_node_id) == false) std::swap(tl_left, tl_right);
+    int sizeof_mask = cat_sets.sizeof_mask(feature_id);
+    // using the odd syntax because += returns post-addition value
+#pragma omp atomic capture
```
Please avoid atomic operations here. If you need to perform a reduction, perform it separately first inside the tree and then across the trees.

This would also store categorical bitmaps from the same tree contiguously.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+    // we have to initialize all pool bytes, so we iterate over those and keep category_it up to
+    // date
+    for (int which_8cats = 0; which_8cats < sizeof_mask; ++which_8cats) {
+      uint8_t _8cats = 0;
```
Please avoid variable names starting with an underscore.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+          ++category_it;
+        }
+      }
+      // bits is a const uint8_t* to issue better load instructions in GPU code
```
This is not a good reason to not to use non-const pointers when building the model. If necessary, define a new type with non-const pointers.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+    }
+    ASSERT(category_it == matching_cats.end(), "internal error: didn't convert all categories");
+  } else
+    ASSERT(false, "only numerical and categorical split nodes are supported");
```
Please add braces around the `else` clause.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -709,6 +779,73 @@ void tl2fil_leaf_payload(fil_node_t* fil_node,
+  } else {
+    node = fil_node_t({}, split, feature_id, default_left, false, is_categorical, fil_left_child);
+  }
+  return {node, tl_left, tl_right};
```
```suggestion
  return conversion_state{node, tl_left, tl_right};
```

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -1061,21 +1224,31 @@ void tl2fil_sparse(std::vector<int>* ptrees,
+  size_t bit_pool_size = 0;
   // convert the nodes
-#pragma omp parallel for
+#pragma omp parallel for shared(bit_pool_size)
```
It's best to avoid using shared variables here.

canonizer commented on cpp/src/fil/fil.cu:
```diff
@@ -1222,13 +1397,28 @@ template <typename threshold_t, typename leaf_t, typename node_t>
   std::stringstream forest_shape = depth_hist_and_max(model);
-  float size_mb =
-    (trees.size() * sizeof(trees.front()) + nodes.size() * sizeof(nodes.front())) / 1e6;
+  float size_mb = (trees.size() * sizeof(trees.front()) + nodes.size() * sizeof(nodes.front()) +
```
```suggestion
  double size_mb = (trees.size() * sizeof(trees.front()) + nodes.size() * sizeof(nodes.front()) +
```

canonizer commented on cpp/src/fil/internal.cuh:
```diff
@@ -38,6 +42,14 @@ __host__ __device__ __forceinline__ int modpow2(int a, int log2_b)
 }
 
+enum adjust_threshold_direction_t { FIL_TO_TREELITE = -1, TREELITE_TO_FIL = 1 };
+void adjust_threshold(float* pthreshold,
```
It's best to keep the threshold adjustment for the implementation and testing separate.

canonizer commented on cpp/src/fil/internal.cuh:
```diff
@@ -141,17 +153,20 @@ __host__ __device__ __forceinline__ val_t base_node::output<val_t>() const
 }
 
 /** dense_node is a single node of a dense forest */
+// template <bool can_be_categorical>
```
Stray commented-out code.

canonizer commented on cpp/src/fil/internal.cuh:
```diff
@@ -141,17 +153,20 @@ __host__ __device__ __forceinline__ val_t base_node::output<val_t>() const
   dense_node(val_t output, val_t split, int fid, bool def_left, bool is_leaf, bool is_categorical)
     : base_node(output, split, fid, def_left, is_leaf, is_categorical)
   {
+    ASSERT(is_categorical == base_node::is_categorical(), "didn't save is_categorical right");
```
This check is clearly unnecessary.

canonizer commented on cpp/src/fil/internal.cuh:
```diff
@@ -141,17 +153,20 @@ __host__ __device__ __forceinline__ val_t base_node::output<val_t>() const
 };
 
 /** sparse_node16 is a 16-byte node in a sparse forest */
+// template <bool can_be_categorical>
```
Stray commented-out code.

canonizer commented on cpp/src/fil/internal.cuh:
```diff
@@ -167,6 +182,7 @@ struct alignas(16) sparse_node16 : base_node {
       left_idx(left_index),
       dummy(0)
   {
+    ASSERT(is_categorical == base_node::is_categorical(), "didn't save is_categorical right");
```
This check is clearly unnecessary.

canonizer commented on cpp/test/sg/fil_test.cu:
```diff
@@ -61,12 +61,12 @@ struct FilTestParams {
   // below, categorical nodes means categorical inner nodes
   // probability that a node is categorical (given that its feature is categorical)
-  float node_categorical_prob = 0.0;
+  float node_categorical_prob = 0.0f;
```
This change clearly does not belong to this pull request.

canonizer commented on cpp/test/sg/fil_test.cu:
```diff
@@ -742,33 +743,39 @@ class TreeliteFilTest : public BaseFilTest {
+          // false
+          threshold = NAN;
+        }
+        adjust_threshold(&threshold, &left_key, &right_key, &default_left, ps.op, FIL_TO_TREELITE);
```
Please avoid using `adjust_threshold` in tests. If necessary, you could implement a separate function for use in tests only.

canonizer commented on cpp/test/sg/fil_test.cu:
```diff
@@ -742,33 +743,39 @@ class TreeliteFilTest : public BaseFilTest {
-          break;
-        default: ASSERT(false, "comparison operator must be <, >, <= or >=");
+      float threshold;
+      if (dense_node.is_categorical()) {
```
Please add treelite test cases for categorical features.

canonizer reviewed (changes requested):
I haven't looked at the Python code yet.

mdoijade committed:
Add hamming, jensen-shannon, kl-divergence, correlation and russellrao distance metrics (#4155)

-- This PR depends on RAFT PR - https://github.com/rapidsai/raft/pull/306
-- Adds cpp & python interfaces for these distance metrics with pytest support for each of them.
-- also remove redundant commented code in canberra distance metric

Authors:
  - Mahesh Doijade (https://github.com/mdoijade)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4155

Nanthini10 committed:
Refactor kmeans sampling code (#4190)

Simplify the type check

Authors:
  - Nanthini (https://github.com/Nanthini10)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4190

Commit:
avoid paramsSolver::{n_rows,n_cols} shadowing their base class counterparts (#4130)

This looks to me like a typo, and may be problematic and confusing if the `n_rows` and `n_cols` members from the base class instead of the ones from the derived class are accessed.

Signed-off-by: Yitao Li <yitao@rstudio.com>

Authors:
  - Yitao Li (https://github.com/yitao-li)

Approvers:
  - Micka (https://github.com/lowener)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4130

levsnv committed:
Apply suggestions from code review

Co-authored-by: Andy Adinets <adinetz@gmail.com>

dantegd committed:
Add thrust_create_target to install export in CMakeLists (#4209)

This will avoid from consumers having to add Thrust explicitly when consuming cuML in CMake. 

cc @shaneding

Authors:
  - Dante Gama Dessavre (https://github.com/dantegd)

Approvers:
  - Robert Maynard (https://github.com/robertmaynard)

URL: https://github.com/rapidsai/cuml/pull/4209

levsnv committed:
FIL to infer categorical forests and generate them in C++ tests (#4092)

This doesn't include treelite import (export). That will come in https://github.com/rapidsai/cuml/pull/4041

Authors:
  - https://github.com/levsnv

Approvers:
  - Andy Adinets (https://github.com/canonizer)
  - Robert Maynard (https://github.com/robertmaynard)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4092

RAMitchell committed:
Implement vector leaf for random forest (#4191)

Fixes #3764,#2518

To do:
- post charts confirming the improvement in accuracy
- address python tests
- benchmark

Authors:
  - Rory Mitchell (https://github.com/RAMitchell)

Approvers:
  - Vinay Deshpande (https://github.com/vinaydes)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4191

levsnv committed:
addressed some review comments

levsnv committed:
added treelite import tests; removed extra vector copy upon feature counter reduction

levsnv committed:
draft change to per-tree bit pool, extra pass instead of an atomic

levsnv committed:
style

levsnv committed:
Merge branch 'categorical-treelite' of github.com:levsnv/cuml into categorical-treelite

levsnv committed:
style

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

cjnolet committed:
Removing experimental flag from hdbscan description in docs (#4211)



Authors:
  - Corey J. Nolet (https://github.com/cjnolet)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4211

venkywonka committed:
RF: python api behaviour refactor (#4207)

This PR ⬇️ 
* fixes #4193 and fixes #4194 that relates to API incompatibility with dask-ml GridSearchCV
* changes the behaviour of cuml RF in the following cases:
    * In the not-so-uncommon case when `n_bins` > number of rows in training sample, instead of throwing error and exiting, the estimator is made to print a warning and use the `n_bins` as the number of training samples. 
    * When `.predict()` is called using `float64` data, instead of throwing an error asking user to explicitly specify `predict_model="CPU"` and rerun, a warning is displayed and implicity defaults to CPU-based prediction from the default GPU-based prediction.
 * Corresponding tests to capture the warnings from above added
 * the estimators now accept both numbers and strings as input for `split_criterion` parameter thus in parity with sklearn's API that takes in strings as criterion.
 * `split_algo` and `use_experimental_backend` parameters of the estimator class have now been completely removed from both documentation and warnings after deprecation in previous releases (from both single-gpu and dask RF). 
 * `num_classes` parameter of predict and score methods have also been similarly removed

Authors:
  - Venkat (https://github.com/venkywonka)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)
  - Rory Mitchell (https://github.com/RAMitchell)

URL: https://github.com/rapidsai/cuml/pull/4207

robertmaynard committed:
Make sure we keep the rapids-cmake and cuml cal version in sync (#4213)

when we make a new cuml version, we need to also bump the rapids-cmake version at the
same time. Otherwise we will get the previous releases dependencies by mistake.

Authors:
  - Robert Maynard (https://github.com/robertmaynard)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4213

levsnv committed:
fixed wrong reduction for max_matching and bad .back() fetch, many prints

levsnv committed:
removed some prints

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

Nyrio committed:
ARIMA - Add support for missing observations and padding (#4058)

This PR allows support for missing observations and padding at the start for variable-length batch. Example:

![missing_obs_0](https://user-images.githubusercontent.com/17441062/125832072-1ff903c9-088e-4d77-9b17-be365890d982.png)

Note: I had to change ARIMA tests because I used a different method than statsmodels (which is used as a reference in tests) to compute the initial parameter estimation. They cut all missing observations for their initial least-square estimation, and I decided to fill them with naive replacements instead, so I keep the temporal relationships in the data and have a much better initial estimate and often a better fit in the end, according to some MASE measurements I made. So I updated the integration test to use the MASE and pass if we are approximately the same _or better_ than statsmodels.

Authors:
  - Louis Sugy (https://github.com/Nyrio)

Approvers:
  - Robert Maynard (https://github.com/robertmaynard)
  - Tamas Bela Feher (https://github.com/tfeher)
  - Dante Gama Dessavre (https://github.com/dantegd)
  - Ray Douglass (https://github.com/raydouglass)

URL: https://github.com/rapidsai/cuml/pull/4058

ajschmidt8 committed:
Merge pull request #4222 from rapidsai/branch-21.08

Forward-merge `branch-21.08` into `branch-21.10`

venkywonka committed:
RF: Add Poisson deviance impurity criterion (#4156)

* Adds the poisson impurity criterion to RF, in parity with scikit learn's RF regressor [[here](https://scikit-learn.org/stable/modules/tree.html#regression-criteria)]
EDIT:
* Also adds C++ level testing for RF Objective function gains of Poisson and Gini.

Authors:
  - Venkat (https://github.com/venkywonka)

Approvers:
  - Rory Mitchell (https://github.com/RAMitchell)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4156

levsnv committed:
style

levsnv committed:
noinline

levsnv committed:
fixed wrong partial_sum application

levsnv committed:
style, restore omp

levsnv committed:
style

levsnv committed:
clean up

hcho3 committed:
Upgrade Treelite to 2.1.0 (#4220)

The 2.1.0 version of Treelite incorporates the following major improvements:

* dmlc/treelite#311
* dmlc/treelite#302
* dmlc/treelite#303
* dmlc/treelite#296

In particular, dmlc/treelite#311 is a critical follow-up to #4191 and addresses a performance regression.

Requires https://github.com/rapidsai/integration/pull/353

Authors:
  - Philip Hyunsu Cho (https://github.com/hcho3)

Approvers:
  - Jordan Jacobelli (https://github.com/Ethyling)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4220

levsnv committed:
Merge branch 'branch-21.10' of github.com:rapidsai/cuml into categorical-treelite

levsnv committed:
removed WAR on exporting empty categorical nodes to treelite

levsnv committed:
noinline

levsnv committed:
...

levsnv committed:
...

levsnv committed:
some review comments, no functional changes

levsnv committed:
Apply suggestions from code review

Co-authored-by: Andy Adinets <adinetz@gmail.com>

divyegala committed:
Using pinned host memory for Random Forest and DBSCAN (#4215)

Benchmarks show that RF performs consistently better with pinned host memory, while DBSCAN sometimes better and sometimes not (within the margin of error), so using pinned host memory by default for both these algorithms.

Ignoring KMeans and LARS for now as both show slightly better perf with pinned host memory but only with increasing number of columns. Since this would need more analysis and deciding if a heuristic is needed for selecting memory, deferring it to 21.12.

Here are the raw numbers:
1. LARS
Normal memory:
```{'lars': {(100000, 10): 0.12429666519165039, (100000, 100): 0.015396833419799805, (100000, 250): 0.015408039093017578, (250000, 10): 0.00986933708190918, (250000, 100): 0.023822546005249023, (250000, 250): 0.03715157508850098, (500000, 10): 0.013423442840576172, (500000, 100): 0.044762372970581055, (500000, 250): 0.07782578468322754}```
Pinned memory:
```{'lars': {(100000, 10): 0.12958097457885742, (100000, 100): 0.01501011848449707, (100000, 250): 0.016597509384155273, (250000, 10): 0.01801013946533203, (250000, 100): 0.022644996643066406, (250000, 250): 0.037090301513671875, (500000, 10): 0.020437955856323242, (500000, 100): 0.044635772705078125, (500000, 250): 0.07696056365966797}```
2. RFR
Normal memory:
```'rfr': {(100000, 10): 1.1951744556427002, (100000, 100): 5.099738359451294, (100000, 250): 11.32804536819458, (250000, 10): 2.0097765922546387, (250000, 100): 9.109776496887207, (250000, 250): 21.058837890625, (500000, 10): 3.3387184143066406, (500000, 100): 15.802990436553955, (500000, 250): 36.80855870246887}```
Pinned memory:
```'rfr': {(100000, 10): 1.1727137565612793, (100000, 100): 4.804195880889893, (100000, 250): 11.621357917785645, (250000, 10): 1.8899295330047607, (250000, 100): 9.16961407661438, (250000, 250): 21.12194561958313, (500000, 10): 3.2937560081481934, (500000, 100): 15.66197681427002, (500000, 250): 36.6080117225647}```
3. KMeans
Normal memory:
```{(100000, 10): 0.11008882522583008, (100000, 100): 0.15475797653198242, (100000, 250): 0.15683507919311523, (250000, 10): 0.18775177001953125, (250000, 100): 0.25696277618408203, (250000, 250): 0.40389132499694824, (500000, 10): 0.4578282833099365, (500000, 100): 0.3917391300201416, (500000, 250): 0.6426849365234375}```
Pinned memory:
```'kmeans': {(100000, 10): 0.11982870101928711, (100000, 100): 0.16992664337158203, (100000, 250): 0.1021108627319336, (250000, 10): 0.16021251678466797, (250000, 100): 0.31025242805480957, (250000, 250): 0.298201322555542, (500000, 10): 0.21084189414978027, (500000, 100): 0.50473952293396, (500000, 250): 0.6191830635070801}```
4. DBSCAN
Normal memory:
```'dbscan': {(100000, 10): 0.4957292079925537, (100000, 100): 0.8680248260498047, (100000, 250): 1.585218906402588, (250000, 10): 4.52524995803833, (250000, 100): 7.175846099853516, (250000, 250): 12.135416269302368, (500000, 10): 26.427770853042603, (500000, 100): 37.57275915145874, (500000, 250): 57.98261737823486}}```
Pinned memory:
```'dbscan': {(100000, 10): 0.49578166007995605, (100000, 100): 0.8678708076477051, (100000, 250): 1.5854766368865967, (250000, 10): 4.526952505111694, (250000, 100): 7.172863006591797, (250000, 250): 12.145166397094727, (500000, 10): 26.422622680664062, (500000, 100): 37.56665277481079, (500000, 250): 58.02563738822937}}```

Authors:
  - Divye Gala (https://github.com/divyegala)

Approvers:
  - Rory Mitchell (https://github.com/RAMitchell)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4215

galipremsagar committed:
Pin max `dask` and `distributed` versions to 2021.09.1 (#4229)

Changes to be in-line with: https://github.com/rapidsai/cudf/pull/9286/

Authors:
  - GALI PREM SAGAR (https://github.com/galipremsagar)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4229

levsnv committed:
"fixed" race condition in n_nodes, python

levsnv committed:
add the empty categorical node workaround back

levsnv committed:
no warning on empty max_matching

levsnv committed:
small review comments

levsnv committed:
small review comments

levsnv committed:
small review comments

levsnv committed:
todo

levsnv committed:
style

levsnv committed:
Merge branch 'categorical-treelite' of github.com:levsnv/cuml into categorical-treelite

levsnv committed:
remove empty category list workaround

achirkin committed:
Faster glm ols-via-eigendecomposition algorithm (#4201)

This PR makes some improvements on glm ordinary least squares (ols) implementation:

  1. Split `MLCommon::LinAlg::lstsq`, which solves OLS by computing SVD of the feature matrix (**UΣV = A**). Originally, this function provided two algorithms to do so (using cusolver QR decomposition and using eigenvalue decomposition).
  2. Add a third algorithm for solving OLS via SVD - via cusolver Jacobi iterations.
  3. Inline `raft::linalg::svdEig`. This function computes SVD decomposition of the feature matrix **UΣV = A** via eigenvalue decomposition of **QΛQ<sup>T</sup> = A<sup>T</sup> A**, which is faster, but requires additional manipulations to recover **Σ** and **U** (while **V = Q**). Instead, I use `raft::linalg::eigDC` directly to compute **w = (A<sup>T</sup> A)<sup>-1</sup>Ab =  (QΛ<sup>-1</sup>Q<sup>T</sup>)(Ab)**. This also allows to compute **Ab** concurrently (in another cuda stream) to the rest of the operations, which often fit in my GPU at the same time when n_cols << n_rows.
  4. Just a small thing: reduce the number of memory allocations using `rmm::device_uvector`. Supposedly, this reduces the wilderness of benchmarking uncertainty when using the default memory allocator.
  5. Force the SVD-Jacobi algorithm when n_cols > n_rows, because none of the others support this case either in theory or due to cusolver limitations.
  6. Update the python interface to allow choosing among all available algorithms.

Authors:
  - Artem M. Chirkin (https://github.com/achirkin)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4201

levsnv committed:
Revert "remove empty category list workaround"

This reverts commit 8ccdb27c49ea2293ee3d9c720cf7822ffa1c5f61.

levsnv committed:
addressed review comments

hcho3 committed:
Expose train_test_split in API doc (#4234)

Currently `train_test_split` isn't showing up in the cuML API doc.

Authors:
  - Philip Hyunsu Cho (https://github.com/hcho3)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4234

mayankanand007 committed:
Adding docs for `.get_feature_names()` inside `TfidfVectorizer` (#4226)

This PR resolves issue #4219 by adding docs for `.get_feature_names()` in the `TfidfVectorizer` class.

As mentioned in the linked issue, the method already exists in `CountVectorizer` and `TfidfVectorizer` inherits from that class, hence the functionality is present but not documented.

Authors:
  - Mayank Anand (https://github.com/mayankanand007)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4226

dantegd committed:
Update scikit-learn version in conda dev envs to 0.24 (#4241)



Authors:
  - Dante Gama Dessavre (https://github.com/dantegd)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)

URL: https://github.com/rapidsai/cuml/pull/4241

levsnv committed:
size_t, uint32_t, int32_t

dantegd committed:
Add option to clone RAFT even if it is in the environment (#4217)

This PR addresses a few issues:

- `libcumlprims` also depends on RAFT so it installs the RAFT headers as part of its own conda package. This made it so that CPM in CI was not cloning RAFT, which made it so that warnings as errors was not being applied to RAFT code, so code that is not covered by RAFT's own C++ tests let warnings trickle in into merged PRs.
- This makes it so that `CPM_raft_source` is always used even if there is a RAFT in the conda include folders.
- Allows users to force CPM to download RAFT even if they have a RAFT in their system already, hopefully alleviating potential folder conflicts that can happen during development

Authors:
  - Dante Gama Dessavre (https://github.com/dantegd)
  - Corey J. Nolet (https://github.com/cjnolet)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4217

dantegd committed:
Merge branch 'branch-21.10' into categorical-treelite
